pkg/controllers/namespace/subnet_share_controller.go

```diff
@@ -0,0 +1,239 @@
+	err := r.Client.Create(ctx, subnetCR)
+	if err != nil {
+		// If the Subnet CR already exists with the same name, try using generateName
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("Subnet CR with the same name already exists, using generateName",
+				"Namespace", subnetCR.Namespace, "Name", subnetName)
+
+			// Create a new Subnet CR with generateName
+			// subnetCR.ObjectMeta.Name will be subnetName + "-" + randomSuffix
+			subnetCR.ObjectMeta.Name = ""
+			subnetCR.ObjectMeta.GenerateName = subnetName + "-"
+
+			err = r.Client.Create(ctx, subnetCR)
+			if err != nil {
+				return fmt.Errorf("failed to create Subnet CR with generateName: %w", err)
+			}
+		} else {
+			log.Error(err, "Failed to create Subnet CR", "Namespace", subnetCR.Namespace, "Name", subnetName)
+			return fmt.Errorf("failed to create Subnet CR: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// createSharedSubnetCR creates a Subnet CR for a shared subnet
+func (r *NamespaceReconciler) createSharedSubnetCR(ctx context.Context, ns string, sharedSubnetPath string) error {
+	// Extract the org id, project id, VPC id, and subnet id
+	orgID, projectID, vpcID, subnetID, err := common.ExtractSubnetPath(sharedSubnetPath)
+	if err != nil {
+		return err
+	}
+
+	projectName, err := r.VPCService.GetProjectName(orgID, projectID)
+	if err != nil {
+		return fmt.Errorf("failed to get project name: %w", err)
+	}
+
+	vpcName, err := r.VPCService.GetVPCName(orgID, projectID, vpcID)
+	if err != nil {
+		return fmt.Errorf("failed to get VPC name: %w", err)
+	}
+
+	// Format VPC full name
+	vpcFullName := fmt.Sprintf("%s:%s", projectName, vpcName)
+	if isDefault, err := r.SubnetService.IsDefaultNSXProject(orgID, projectID); err != nil {
+		return fmt.Errorf("failed to check if project is default: %w", err)
+	} else if isDefault {
+		vpcFullName = fmt.Sprintf(":%s", vpcName)
+	}
+
+	// Get associated resource name
+	associatedName, err := common.ConvertSubnetPathToAssociatedResource(sharedSubnetPath)
+	if err != nil {
+		return err
+	}
+
+	// Create the Subnet CR object
+	subnetCR := r.SubnetService.BuildSubnetCR(ns, subnetID, vpcFullName, associatedName)
+
+	// Create the Subnet CR in Kubernetes
+	err = r.createSubnetCRInK8s(ctx, subnetCR, subnetID)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Created Subnet CR for shared Subnet", "Namespace", ns, "Name", subnetCR.Name, "SharedSubnet", sharedSubnetPath)
+	return nil
+}
+
+// getExistingSharedSubnetCRs gets a map of existing shared Subnet CRs in the namespace
+func (r *NamespaceReconciler) getExistingSharedSubnetCRs(ctx context.Context, ns string) (map[string]*v1alpha1.Subnet, error) {
+	// Get the list of Subnet CRs in the namespace that have the associated_resource annotation
+	subnetList := &v1alpha1.SubnetList{}
+	err := r.Client.List(ctx, subnetList,
+		client.InNamespace(ns),
+		client.MatchingFields{
+			"metadata.annotations." + servicecommon.AnnotationAssociatedResource: "true",
```

Comment:
You may need to register the index function in Namespace controller to use this?

Reply:
Field indexers are registered at the **Manager level**, not at the controller level. Once set up a field indexer through any controller, **all controllers using the same manager can use that index**